### PR TITLE
Apply non-strict rules for PT appliance on project import to avoid wo…

### PR DESCRIPTION
**Pull request by gazarenkov**
@vparfonov @azatsarynnyy  please review. 
### What does this PR do?

Improves solving the PT related problems with importing Project. Such as: 
- The System (Che or Codenvy) does not contain Project Types described in serialized Project configuration. 
-  There are non-conformance between PT descriptions so serialized Project configuration  does not contain required (as for PTs registered in the System) Attributes
### What issues does this PR fix or reference?

none
### Previous behavior

Initially import of such data broke Workspace creation.
After it was some improved so ANY problem caused adding Problem record (only one) with description replacing primary PT with Blank, removing all the mixins and attributes.
### New behavior

In a case of problematic import:
- Added as many Problem record as needed (all the problems are covered)
- In a case of problem with primary PT replaces it with Blank and removes only Attributes belonged to initial PT
- In a case of problem with mixin PT removes only it and removes only Attributes belonged to initial mixin PT

Please review [Che's Contributing Guide](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md) for best practices.

…rkspace failing

### CODE REVIEW

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 1]
```diff
@@ -30,32 +33,45 @@
...
 
+        ProjectTypeDef tmpPrimary;
         if (type == null) {
-            throw new ProjectTypeConstraintException("No primary type defined for " + projectPath);
-        }
 
-        primary = projectTypeRegistry.getProjectType(type);
+            problems.add(new Problem(12, "No primary type defined for " + projectPath + " Base Project Type assigned."));
+            tmpPrimary = ProjectTypeRegistry.BASE_TYPE;
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: remove extra empty line.

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 2]
```diff
@@ -30,32 +33,45 @@
...
+        } else {
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: remove extra empty line.

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 3]
```diff
@@ -116,6 +142,50 @@ public ProjectTypeDef getPrimary() {
...
+     * in case when some attributes are not valid
+     * (for instance required attributes are not initialized)
+     *
+     * @param attributesToDel - invalid attributes
+     */
+    void reset(Set<Attribute> attributesToDel) {
+
+        Set<String> ptsToDel = new HashSet<>();
+        for(Attribute attr : attributesToDel) {
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: remove extra empty line.

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java [hunk 4]
```diff
@@ -250,6 +236,19 @@ public FolderEntry getBaseFolder() {
     }
 
     /**
+     * @return list of Problems as a String
+     */
+    public String getProblemsStr() {
+        String problemsStr = "";
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: StringBuilder or http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Joiner.html?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 5]
```diff
@@ -116,6 +142,50 @@ public ProjectTypeDef getPrimary() {
...
+                mixins.remove(typeId);
+            }
+        }
+
+        // remove attributes
+        for(String attr : attrNamesToDel) {
+            this.attributeDefs.remove(attr);
+        }
+
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: remove extra empty line.

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java [hunk 6]
```diff
@@ -75,8 +75,18 @@
         problems = new ArrayList<>();
         attributes = new HashMap<>();
 
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: remove extra empty line

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java [hunk 7]
```diff
@@ -75,8 +75,18 @@
...
+            path = folder.getPath();
+        } else if (config != null) {
+            path = Path.of(config.getPath());
+        } else {
+            throw new ServerException("Invalid Project Configuration. Path undefined.");
+        }
+
         this.folder = folder;
-        this.config = (config == null) ? new NewProjectConfig(folder.getPath()) : config;
+        this.config = (config == null) ? new NewProjectConfig(/*folder.getPath()*/ path) : config;
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: strange comment inside of constructor

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java (hunk 1 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: ok

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java (hunk 2 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: ok

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java (hunk 5 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: ok

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java (hunk 6 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: ok

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java (hunk 7 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: fixed

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java (hunk 4 above)
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java]: ok

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 8]
```diff
@@ -116,6 +142,50 @@ public ProjectTypeDef getPrimary() {
...
+            ProjectTypeDef typeDef = all.get(pt);
+            for(Attribute attrDef: typeDef.getAttributes()) {
+                attrNamesToDel.add(attrDef.getName());
+            }
+        }
+
+        // remove project types
+        for(String typeId : ptsToDel) {
+
```
[the remark this answers is not in the document]
Comment (gazarenkov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java]: ok

**Review by vparfonov — APPROVED** [review]

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 9]
```diff
@@ -264,6 +264,17 @@ public RegisteredProject createProject(ProjectConfig projectConfig, Map<String,
                 throw e;
             }
 
+            // unlike imported it is not appropriate for newly created project to have problems
+            if(!project.getProblems().isEmpty()) {
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:268]: invalid format :  space is missing before left parenthesis

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 10]
```diff
@@ -308,8 +319,18 @@ public RegisteredProject updateProject(ProjectConfig newConfig) throws Forbidden
             throw new NotFoundException(String.format("Folder '%s' doesn't exist.", path));
         }
 
+        ProjectConfig oldConfig = projectRegistry.getProject(path);
+
         final RegisteredProject project = projectRegistry.putProject(newConfig, baseFolder, true, false);
 
+        // unlike imported it is not appropriate for updated project to have problems
+        if(!project.getProblems().isEmpty()) {
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:327]: wrong code style for if

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 11]
```diff
@@ -320,6 +341,23 @@ public RegisteredProject updateProject(ProjectConfig newConfig) throws Forbidden
...
+     *
+     * Import source code as a Basic type of Project
+     *
+     * @param path where to import
+     * @param sourceStorage where sources live
+     * @param rewrite whether rewrite or not (throw exception othervise) if such a project exists
+     *
+     * @return Project
+     *
+     * @throws ServerException
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:354]: exceptions should be written: what is the cause of all of them
these format is useless

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 12]
```diff
@@ -379,12 +417,23 @@ public RegisteredProject importProject(String path, SourceStorage sourceStorage,
...
+    /**
+     * Estimates if the folder can be treated as a project of particular type
+     *
+     * @param path to the folder
+     * @param projectTypeId project type to estimate
+     *
+     * @return resolution object
+     * @throws ServerException
+     * @throws NotFoundException
+     * @throws ValueStorageException
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:429]: be more explicit on exceptions thrown

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 13]
```diff
@@ -396,7 +445,16 @@ public ProjectTypeResolution estimateProject(String path, String projectTypeId)
...
     }
 
-    // ProjectSuggestion
+    /**
+     * Estimates to which project types the folder can be converted to
+     *
+     * @param path to the folder
+     * @param transientOnly whether it can be estimated to the transient types of Project only
+     *
+     * @return list of resolutions
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:454]: be explicit on exceptions

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 14]
```diff
@@ -443,6 +501,20 @@ public void delete(String path) throws ServerException, ForbiddenException, NotF
         workspaceProjectsHolder.sync(projectRegistry);
     }
 
+    /**
+     * Copies item to new path with
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:505]: with what ?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 15]
```diff
@@ -478,6 +550,20 @@ public VirtualFileEntry copyTo(String itemPath, String newParentPath, String new
...
 
+    /**
+     * Moves item to the new path
+     *
+     * @param itemPath path to the item
+     * @param newParentPath path of new parent
+     * @param newName new item's name
+     * @param overwrite whether existed (if any) item should be overwritten
+     *
+     * @return new item
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:561]: missing {@link }

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 16]
```diff
@@ -478,6 +550,20 @@ public VirtualFileEntry copyTo(String itemPath, String newParentPath, String new
...
+     * @throws ServerException
+     * @throws NotFoundException
+     * @throws ConflictException
+     * @throws ForbiddenException
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:565]: what are these exceptions about ?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 17]
```diff
@@ -30,32 +33,42 @@
...
         this.projectTypeRegistry = projectTypeRegistry;
         this.projectPath = projectPath;
 
+        ProjectTypeDef tmpPrimary;
         if (type == null) {
-            throw new ProjectTypeConstraintException("No primary type defined for " + projectPath);
-        }
-
-        primary = projectTypeRegistry.getProjectType(type);
+            problems.add(new Problem(12, "No primary type defined for " + projectPath + " Base Project Type assigned."));
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:55]: what is 12 value ? it should be a constant

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 18]
```diff
@@ -30,32 +33,42 @@
...
+            tmpPrimary = ProjectTypeRegistry.BASE_TYPE;
+        } else {
+            try {
+                tmpPrimary = projectTypeRegistry.getProjectType(type);
+            } catch (NotFoundException e) {
+                problems.add(new Problem(12, "Primary type " + type + " defined for " + projectPath + " is not registered. Base Project Type assigned."));
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:61]: 12 --> constant

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 19]
```diff
@@ -73,24 +86,32 @@ public ProjectTypes(String projectPath,
...
+            try {
+                mixin = projectTypeRegistry.getProjectType(mixinFromConfig);
+            } catch (NotFoundException e) {
+                problems.add(new Problem(12, "Project type " + mixinFromConfig + " is not registered. Skipped."));
+                continue;
+            }
+
             if (!mixin.isMixable()) {
-                throw new ProjectTypeConstraintException("Project type " + mixin + " is not allowable to be mixin. It not mixable");
+                problems.add(new Problem(12, "Project type " + mixin + " is not allowable to be mixin. It not mixable. Skipped."));
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:98]: 12 = constant

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 20]
```diff
@@ -73,24 +86,32 @@ public ProjectTypes(String projectPath,
...
             for (Attribute attr : mixin.getAttributes()) {
                 if (attributeDefs.containsKey(attr.getName())) {
-                    throw new ProjectTypeConstraintException(
-                            "Attribute name conflict. Duplicated attributes detected " + projectPath +
-                            " Attribute " + attr.getName() + " declared in " + mixin.getId() + " already declared in " +
-                            attributeDefs.get(attr.getName()).getProjectType()
-                    );
-                }
 
+                    problems.add(new Problem(13, "Attribute name conflict. Duplicated attributes detected " + projectPath +
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:110]: 13 = ? constant

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 21]
```diff
@@ -116,6 +137,45 @@ public ProjectTypeDef getPrimary() {
...
+    /**
+     * Reset project types and atrributes after initialization
+     * in case when some attributes are not valid
+     * (for instance required attributes are not initialized)
+     *
+     * @param attributesToDel - invalid attributes
+     */
+    void reset(Set<Attribute> attributesToDel) {
+
+        Set<String> ptsToDel = new HashSet<>();
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:149]: size is known ? should be set

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java [hunk 22]
```diff
@@ -174,14 +154,19 @@ private void initAttributes() throws ValueStorageException, ProjectTypeConstrain
                 }
 
                 if (value.isEmpty() && variable.isRequired()) {
-                    throw new ProjectTypeConstraintException("Value for required attribute is not initialized " + variable.getId());
+                    this.problems.add(new Problem(13, "Value for required attribute is not initialized " + variable.getId()));
+                    invalidAttributes.add(variable);
+                    //throw new ProjectTypeConstraintException("Value for required attribute is not initialized " + variable.getId());
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/RegisteredProject.java:159]: comment should be removed  or it's a TODO or FIXME

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 23]
```diff
@@ -116,6 +137,45 @@ public ProjectTypeDef getPrimary() {
...
+        for(Attribute attr : attributesToDel) {
+            ptsToDel.add(attr.getProjectType());
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:151]: should use new Java8 loops

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java [hunk 24]
```diff
@@ -116,6 +137,45 @@ public ProjectTypeDef getPrimary() {
...
+        }
+
+        Set <String>attrNamesToDel = new HashSet<>();
+        for(String pt : ptsToDel) {
```
Comment (benoitf) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectTypes.java:155]: should be reworked with Java8 loops

**Pull request merged**